e3sm_diags/e3sm_diags_driver.py

```diff
@@ -256,92 +240,222 @@ def get_parameters(parser=CoreParser()):
     return parameters
 
 
-def run_diag(parameters):
-    """
-    For a single set of parameters, run the corresponding diags.
+def create_parameter_dict(parameters):
+    d: Dict[type, int] = dict()
+    for parameter in parameters:
+        t = type(parameter)
+        if t in d.keys():
+            d[t] += 1
+        else:
+            d[t] = 1
+    return d
+
+
+def run_diag(parameter: CoreParameter) -> List[CoreParameter]:
+    """Run the corresponding diagnostics for a set of parameters.
+
+    Additional CoreParameter (or CoreParameter sub-class) objects are derived
+    from the CoreParameter `sets` attribute, hence this function returns a
+    list of CoreParameter objects.
+
+    This function loops over the specified diagnostic sets and runs the
+    diagnostic using the parameters.
+
+    Parameters
+    ----------
+    parameter : CoreParameter
+        The CoreParameter object to run diagnostics on.
+
+    Returns
+    -------
+    List[CoreParameter]
+        The list of CoreParameter objects with results from the diagnostic run.
     """
     results = []
-    for set_name in parameters.sets:
 
-        parameters.current_set = set_name
+    for set_name in parameter.sets:
+        # FIXME: The parameter and driver for a diagnostic should be mapped
+        # together. If this is done, the `run_diag` function should be
+        # accessible by the parameter class without needing to perform a static
+        # string reference for the module name.
+        parameter.current_set = set_name
         mod_str = "e3sm_diags.driver.{}_driver".format(set_name)
+
+        # Check if there is a matching driver module for the `set_name`.
         try:
             module = importlib.import_module(mod_str)
-            single_result = module.run_diag(parameters)
-            print("")
-            results.append(single_result)
-        except Exception:
-            logger.exception("Error in {}".format(mod_str), exc_info=True)
-            traceback.print_exc()
-            if parameters.debug:
-                sys.exit()
+        except ModuleNotFoundError as e:
+            logger.error(f"'Error with set name {set_name}'", exc_info=e)
+            continue
+
+        # Call the driver module's `run_diag` function.
+        if module is not None:
+            try:
+                single_result = module.run_diag(parameter)
+                results.append(single_result)
+            except Exception:
+                logger.exception(f"Error in {mod_str}", exc_info=True)
+
+                if parameter.debug:
+                    sys.exit()
```

Comment:
The original `try` and `except` statement did not catch when a driver module was not found for a `set_name`.

To handle this, I refactored this logic to split getting the driver module and then calling `run_diag`.